invenio_vocabularies/factories.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 CERN.
+# Copyright (C) 2024 KTH Royal Institute of Technology.
+#
+# Invenio-Vocabularies is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+"""Factory for Vocabulary Config."""
+from copy import deepcopy
+
+import yaml
+from invenio_records_resources.proxies import current_service_registry
+
+from .contrib.awards.datastreams import DATASTREAM_CONFIG as awards_ds_config
+from .contrib.funders.datastreams import DATASTREAM_CONFIG as funders_ds_config
+from .contrib.names.datastreams import DATASTREAM_CONFIG as names_ds_config
+
+
+class VocabularyConfigFactory:
+    """Base class for Vocabulary Config Factory."""
+
+    def get_config(self, filepath=None, origin=None):
```

Comment:
so third try to create that review:

since the get_config is the same in all three sub classes i would move the code to the base class `VocabularyConfig` and do customization in the sub classes by adding a `vocabulary_name` and `config` property.

```
class VocabularyConfig:
    config = None
    vocabulary_name = None

    def get_config(self, filepath=None, origin=None):
        """Get the configuration for the names vocabulary."""
        # maybe add some checks to help further self to use the subclasses in the right way
        config = deepcopy(self.config)
        if filepath:
            with open(filepath, encoding="utf-8") as f:
                config = yaml.safe_load(f).get(self.vocabulary_name)
        if origin:
            config["readers"][0].setdefault("args", {})
            config["readers"][0]["args"]["origin"] = origin
        return config

    def get_service(self):
        """Get the service for the names vocabulary."""
        return current_service_registry.get(self.vocabulary_name)

class NamesVocabularyConfig:
    config = names_ds_config
    vocabulary_name = "names"
```
i have removed `Factory` from the names, because in my opinion those config classes are not creating anything and are therefore no factory classes

Reply:
@utnapischtim After taking another look, you're absolutely right—thank you for highlighting that. I was probably swayed by the comment on the Fixme...
But if it's not a factory, should we relocate the code back to `cli.py`, `config.py` or is it best to leave it where it currently is?

Reply:
btw in the original code only the "names" had a service. Now all vocabs could return a service. Could that be a problem?

Reply:
Good point. We can raise a `NotImplementedError` in the get_service methods for both `funders` and `awards` in that case. Unless you have another approach?

Reply:
we could do something like:

```
class VocabularyConfig:
    service_name = None
    
    def get_service(self):
        if self.service_name is None:
            raise NotImplementedError
        return current_service_registry.get(self.service_name)
```
and set the name of the service only in NamesVocabularyConfig

Reply:
I've opted to raise `NotImplementedError` in the subclass get_service methods for  `funders` and `awards`. 
This way, it's unambiguous that these services aren't available yet. Once implemented, we can simply remove the override to enable the service.